### PR TITLE
Feat: setup reserved workspaces when calling list workspaces API

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/server/workspace_client.ts [hunk 1]
```diff
@@ -125,6 +227,52 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
           type: WORKSPACE_TYPE,
         }
       );
+      const scopedClientWithoutPermissionCheck = this.getScopeClientWithoutPermisson(requestDetail);
```
Comment (yuye-aws) [inline src/plugins/workspace/server/workspace_client.ts:230]: There has been a PR fixing typos, please change the function name to `getScopedClientWithoutPermission`.

File: src/plugins/workspace/server/workspace_client.ts [hunk 2]
```diff
@@ -65,6 +79,94 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
...
+        features: ['*', `!@${DEFAULT_APP_CATEGORIES.management.id}`],
+        reserved: true,
+      },
+      publicWorkspaceACL.getPermissions()
+    );
+  }
+  private async setupManagementWorkspace(savedObjectClient?: SavedObjectsClientContract) {
+    const managementWorkspaceACL = new ACL().addPermission([WorkspacePermissionMode.Management], {
+      users: ['*'],
+    });
```
Comment (yuye-aws) [inline src/plugins/workspace/server/workspace_client.ts:126]: For management workspace, are we going to grant `WorkspacePermissionMode.Management` to all users?

File: src/plugins/workspace/server/workspace_client.ts [hunk 3]
```diff
@@ -65,6 +79,94 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
...
+        }
+      } catch (e) {
+        this.logger.error(`Create ${workspaceId} workspace error: ${e?.toString() || ''}`);
+      }
+    }
+  }
+  private async setupPublicWorkspace(savedObjectClient?: SavedObjectsClientContract) {
+    const publicWorkspaceACL = new ACL().addPermission([WorkspacePermissionMode.Management], {
+      users: ['*'],
+    });
```
Comment (yuye-aws) [inline src/plugins/workspace/server/workspace_client.ts:109]: For public workspace, are we going to grant `WorkspacePermissionMode.Management` to all users?

**Comment by yuye-aws** [comment]
I found that reserved workspaces were not automatically created from the left menu header. When workspace feature is enabled, the header obtain the workspace list like the following: 
```
src/plugins/workspace/public/components/workspace_menu/workspace_menu.tsx
const workspaceList = useObservable(observables.workspaceList$, []);
```
Instead of simply get `workspaceList$` observable, we need to call `workspace _list` to make this PR effective.

File: src/plugins/workspace/server/workspace_client.ts (hunk 2 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts:126]: Yes, the conclusion is that we will make public and management workspace manageable by anyone, and the first one can change the permission if they want to limit the permission.

File: src/plugins/workspace/server/workspace_client.ts (hunk 3 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts:109]: Yes, the conclusion is that we will make public and management workspace manageable by anyone, and the first one can change the permission if they want to limit the permission.

**Comment by SuZhou-Joe** [comment]
> I found that reserved workspaces were not automatically created from the left menu header. When workspace feature is enabled, the header obtain the workspace list like the following:
> 
> ```
> src/plugins/workspace/public/components/workspace_menu/workspace_menu.tsx
> const workspaceList = useObservable(observables.workspaceList$, []);
> ```
> 
> Instead of simply get `workspaceList$` observable, we need to call `workspace _list` to make this PR effective.

There should be a place to call the workspace _list or the workspaceList$ will always be empty.

File: src/plugins/workspace/server/workspace_client.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts]: Fixed and thanks

File: src/plugins/workspace/server/workspace_client.ts [hunk 4]
```diff
@@ -125,6 +227,52 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
...
+       */
+      const principals = SavedObjectsPermissionControl.getPrincipalsFromRequest(
+        requestDetail.request
+      );
+      /**
+       * Only when authentication is enabled will personal workspace be created
+       */
+      if (principals.users) {
+        const hasPersonalWorkspace = savedObjects.find((item) =>
+          principals.users?.includes(item.id)
```
Comment (ruanyl) [inline src/plugins/workspace/server/workspace_client.ts:263]: Can you leave a comment and say something like: user id will be personal workspace id

File: src/plugins/workspace/server/workspace_client.ts [hunk 5]
```diff
@@ -125,6 +227,52 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
           type: WORKSPACE_TYPE,
         }
       );
+      const scopedClientWithoutPermissionCheck = this.getScopedClientWithoutPermission(requestDetail);
```
Comment (yuye-aws) [inline src/plugins/workspace/server/workspace_client.ts:230]: Please fix lint

File: src/plugins/workspace/server/permission_control/client.ts [hunk 6]
```diff
@@ -29,19 +29,7 @@ export interface AuthInfo {
...
-      excludedWrappers: [WORKSPACE_SAVED_OBJECTS_CLIENT_WRAPPER_ID],
-    });
-  }
-
-  constructor(logger: Logger) {
-    this.logger = logger;
-  }
-
-  public getPrincipalsFromRequest(request: OpenSearchDashboardsRequest): Principals {
+  static getPrincipalsFromRequest(request: OpenSearchDashboardsRequest): Principals {
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/permission_control/client.ts]: Agree, changed that.

File: src/plugins/workspace/server/workspace_client.ts [hunk 7]
```diff
@@ -65,6 +79,94 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
...
+  private async setupPublicWorkspace(savedObjectClient?: SavedObjectsClientContract) {
+    const publicWorkspaceACL = new ACL().addPermission([WorkspacePermissionMode.Management], {
+      users: ['*'],
+    });
+    return this.checkAndCreateWorkspace(
+      savedObjectClient,
+      PUBLIC_WORKSPACE_ID,
+      {
+        name: i18n.translate('workspaces.public.workspace.default.name', {
+          defaultMessage: 'public',
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts:116]: Thanks for this catch. According to mockup, it should be named "Global workspace", changed to that.

File: src/plugins/workspace/server/workspace_client.ts [hunk 8]
```diff
@@ -125,6 +227,52 @@ export class WorkspaceClientWithSavedObject implements IWorkspaceDBImpl {
           type: WORKSPACE_TYPE,
         }
       );
+      const scopedClientWithoutPermissionCheck = this.getScopedClientWithoutPermission(requestDetail);
+      const tasks: Array<Promise<unknown>> = [];
+
+      /**
+       * Setup public workspace if public workspace can not be found
+       */
+      const hasPublicWorkspace = savedObjects.find((item) => item.id === PUBLIC_WORKSPACE_ID);
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts:238]: Sure, changed that.

File: src/plugins/workspace/server/workspace_client.ts (hunk 4 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts]: Sure, done.

**Comment by raintygao** [comment]
Will permission related changes in this PR be conflicted with this task? https://github.com/orgs/opensearch-project/projects/127/views/5?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=Assignees&pane=issue&itemId=38846476

**Comment by SuZhou-Joe** [comment]
> Will permission related changes in this PR be conflicted with this task? https://github.com/orgs/opensearch-project/projects/127/views/5?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=Assignees&pane=issue&itemId=38846476

Let me change the setup permission to comply with the new permission model.

**Review by ruanyl — APPROVED** [review]

**Review by yuye-aws — APPROVED** [review]

**Review by raintygao — APPROVED** [review]

**Pull request merged**